File: erpnext/controllers/transaction.py
# ERPNext - web based ERP (http://erpnext.com)
# Copyright (C) 2012 Web Notes Technologies Pvt Ltd
# 
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

from __future__ import unicode_literals
import webnotes
import webnotes.model
from webnotes.model import get_controller
from webnotes.model.controller import DocListController

class TrasactionController(DocListController):
	def validate_items_life(self):
		for child in self.doclist.get({"parentfield": self.item_table_fieldname}):
			get_controller("Item", child.item_code).check_end_of_life()
	
	# def validate_previous_doclist(self, fieldname, item_fieldname):
	# 	"""Validates current item row against previous item row"""
	# 	doctypelist = webnotes.model.get_doctype(self.doc.doctype)
	# 	prev_doctype = doctypelist.get_field(fieldname,
	# 	 	parentfield=self.item_table_fieldname).options
	# 
	# 	def errmsg(errfield, item, prev_item):
	# 		label = doctypelist.get_label(errfield, parent=item.doctype)
	# 		webnotes.msgprint("""%s of row # %d in the %s table
	# 			cannot be different from the one in row # %d in the %s table 
	# 			of %s: "%s" """ % (label, item.idx, item.doctype, prev_item.idx, 
	# 			prev_item.doctype, prev_doctype, item.get(fieldname)),
	# 			raise_exception=True)
	# 
	# 	prev_doclist_cache = {}
	# 	for item in self.doclist.get({"parentfield": self.item_table_fieldname}):
	# 		if item.get(fieldname):
	# 			prev_item = prev_doclist_cache.setdefault(item.get(fieldname),
	# 				webnotes.model.get(prev_doctype, name)).getone({
	# 					"name": item.get(item_fieldname)})
	# 			
	# 			if item.item_code != prev_item.item_code:
	# 				# mismatch in item code
	# 				errmsg("item_code", item, prev_item)
	# 			if item.warehouse != prev_item.warehouse:
	# 				# mismatch in warehouse
	# 				errmsg("warehouse", item, prev_item)
	# 			if item.uom != prev_item.uom:
	# 				# mismatch in UOM
	# 				errmsg("uom", item, prev_item)

File: public/js/transaction.js
// ERPNext - web based ERP (http://erpnext.com)
// Copyright (C) 2012 Web Notes Technologies Pvt Ltd
// 
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

wn.provide("erpnext");

erpnext.Transaction = Class.extend({
	init: function(opts) {
		$.extend(this, opts);
		this.setup_get_query && this.setup_get_query();
		this.load_precision_maps && this.load_precision_maps();
	},
	
	onload: function() {
		this.set_missing_values();
	},
	
	refresh: function() {
		erpnext.hide_naming_series();
		if(this.add_buttons) {
			this.frm.clear_custom_buttons();
			this.add_buttons();
		}
		this.toggle_fields && this.toggle_fields();
	},
	
	set_missing_values: function() {
		var me = this;
		var default_values = {
			posting_date: dateutil.obj_to_str(new Date()),
			status: "Draft",
			company: sys_defaults.company,
			fiscal_year: sys_defaults.fiscal_year,
			currency: sys_defaults.currency,
			exchange_rate: 1.0,
			
			// for buying
			is_subcontracted: "No",
			
			// for selling
			price_list_name: sys_defaults.price_list_name,
			price_list_currency: sys_defaults.price_list_currency,
			plc_exchange_rate: 1,
		}
		$.each(default_values, function(key, value) {
			if(!me.frm.doc[key] && wn.meta.get_docfield(me.frm.doc.doctype, key)) {
				me.frm.doc[key] = value;
			}
		});
	},
	
	item_code: function(doc, cdt, cdn) {
		var me = this;
		var item = locals[cdt][cdn];
		if(item.item_code) {
			wn.call({
				doc: me.frm.doc,
				method: "get_item_details",
				args: {
					item_code: item.item_code,
					warehouse: item.warehouse,
					income_account: item.income_account,
					expense_account: item.expense_account,
					cost_center: item.cost_center
				},
				callback: function() {
					// update item doc
					$.extend(locals[cdt][cdn], r.message);
					if(this.custom_item_code) {
						this.custom_item_code(doc, cdt, cdn);
					}
					refresh_field(me.item_table_field);
				}
			});
		}
	},

	barcode: function(doc, cdt, cdn) {
		var me = this;
		var item = locals[cdt][cdn];
		if (item.barcode) {
			wn.call({
				doc: me.frm.doc,
				method: "get_barcode_details",
				args: {
					barcode: item.barcode,
					warehouse: item.warehouse,
					income_account: item.income_account,
					expense_account: item.expense_account,
					cost_center: item.cost_center
				}
				callback: function(r, rt) {
					// update item doc
					$.extend(locals[cdt][cdn], r.message);
					refresh_field(me.item_table_field);
				}
			});
		}
	},

	setup_get_query: function() {
		var me = this;
		
		// taxes and charges master
		if(this.frm.fields_dict.taxes_and_charges_master) {
			this.frm.fields_dict.taxes_and_charges_master.get_query = function() {
				return repl("select distinct name from `tab%(doctype)s` \
					where company = \"%(company)s\" and docstatus < 2 \
					and %(key)s like \"%s\" order by name limit 50", {
						doctype: wn.meta.get_docfield(me.frm.doc.doctype,
							"taxes_and_charges_master").options.split("\n")[0],
						company: me.frm.doc.company
					});
			}
		}
	},
	
	get_address: function(args) {
		wn.call({
			doc: this.frm.doc,
			method: 'get_address',
			args: args || {},
			callback: function(r) {
				me.frm.refresh();
			}
		});
	},
	
	get_contact: function(args) {
		wn.call({
			doc: this.frm.doc,
			method: 'get_contact',
			args: args || {},
			callback: function(r) {
				me.frm.refresh();
			}
		});
	},
	
});

erpnext.InputDialog = function(title, input_label, method) {
	var me = this;
	dialog = new Dialog(400, 400, title);
	dialog.make_body([
		['Text', input_label],
		['HTML', 'response_area', '<div class = "comment" id="response"></div>'],
		['HTML', 'button_area', '<div></div>']
	]);
	var add_button = $a($i(dialog.widgets['button_area']), 'button', 'button');
	add_button.innerHTML = 'Add';
	add_button.onclick = function() { dialog.add(); }

	var cancel_button = $a($i(dialog.widgets['button_area']), 'button', 'button');
	cancel_button.innerHTML = 'Cancel';
	$y(cancel_button, { marginLeft: '4px' });
	cancel_button.onclick = function() { dialog.hide(); };
	
	dialog.onshow = function() {
		dialog.widgets[input_label].value = '';
		$i('response').innerHTML = '';
	}

	dialog.add = function() {
		$i('response').innerHTML = 'Processing...';
		var arg = strip(dialog.widgets[input_label].value);
		if(arg) {
			wn.call({
				method: "runserverobj",
				args: {
					docs: wn.model.compress(wn.model.get_doclist(me.frm.doc.doctype,
						me.frm.doc.name)),
					method: method,
					args: arg
				},
				callback: function(r, rt) {
					if(r.message) {
						$i('response').innerHTML = 'Done';
						dialog.hide();
					}
				}
			});
		} else {
			$i('response').innerHTML = 'Please enter somthing before adding';
		}
	}
}
